src/mlpack/methods/ann/layer/batch_norm_impl.hpp

```diff
@@ -55,64 +55,64 @@ template<typename eT>
 void BatchNorm<InputDataType, OutputDataType>::Forward(
     const arma::Mat<eT>&& input, arma::Mat<eT>&& output)
 {
-  output.reshape(input.n_rows, input.n_cols);
-
   // Mean and variance over the entire training set will be used to compute
   // the forward pass when deterministic is set to true.
   if (deterministic)
   {
     mean = stats.mean();
     variance = stats.var(1);
+
+    output = input.each_col() - mean;
+    output.each_col() %= gamma / arma::sqrt(variance + eps);
+    output.each_col() += beta;
   }
   else
   {
     mean = arma::mean(input, 1);
     variance = arma::var(input, 1, 1);
 
-    for (size_t i = 0; i < output.n_cols; i++)
-    {
+    for (size_t i = 0; i < input.n_cols; i++)
       stats(input.col(i));
-    }
-  }
 
-  output = input.each_col() - mean;
-  output.each_col() %= gamma / arma::sqrt(variance + eps);
-  output.each_col() += beta;
+    output = input.each_col() - mean;
+    output.each_col() /= arma::sqrt(variance + eps);
+
+    // Reused in the backward and gradient step.
+    normalized = output;
+
+    output.each_col() %= gamma;
+    output.each_col() += beta;
+  }
 }
 
 template<typename InputDataType, typename OutputDataType>
 template<typename eT>
 void BatchNorm<InputDataType, OutputDataType>::Backward(
     const arma::Mat<eT>&& input, arma::Mat<eT>&& gy, arma::Mat<eT>&& g)
 {
-  mean = arma::mean(input, 1);
-  variance = arma::var(input, 1, 1);
-
-  arma::mat m = arma::sum(gy % (input.each_col() - mean), 1);
-  g = (mean - input.each_col());
-  g.each_col() %= m;
-  g.each_col() %= 1.0/(variance + eps);
-  g += (gy.each_col() - arma::sum(gy, 1));
-  g += (input.n_cols - 1) * gy;
-  g.each_col() %= ((1.0 / input.n_cols) * gamma);
-  g.each_col() %= (1.0 / arma::sqrt(variance + eps));
+  const arma::mat inputMean = input.each_col() - mean;
+  const arma::mat stdInv = 1.0 / arma::sqrt(variance + eps);
+  const arma::mat norm = gy.each_col() % gamma;
+  const arma::mat var = arma::sum(norm % inputMean, 1) %
+      arma::pow(stdInv, 3.0) * 0.5;
+
+  g = (norm.each_col() % stdInv) + (inputMean.each_col() %
+      var * 2 / input.n_cols);
+  g.each_col() += arma::sum(norm.each_col() % -stdInv, 1) + var %
+      arma::mean(-2 * inputMean, 1) / input.n_cols;
```

Comment:
Do you think it would be worthwhile to rewrite this expression to avoid the temporaries `inputMean`, `stdInv`, `norm`, and `var`, similar to how it was written before?

Reply:
Sorry, I spent a lot more time with this this morning and I think that my comment above is misguided---since some of these variables are used multiple times, it doesn't make sense to avoid the temporaries.

I used the paper to try and understand this code as the equation for ∂l/∂xᵢ on page 4, but I am a little bit confused.  I can see that:

 * `(norm.each_col() % stdInv)` corresponds directly to (∂l/∂x̂) * (1 / sqrt(σ_B^2 + ϵ)).
 * `(inputMean.each_col() % var * 2 / input.n_cols)` corresponds directly to (∂l/∂σ_B^2) * 2 (xᵢ - µ_B) / m.
 * `(arma::sum(norm.each_col() % -stdInv, 1))` corresponds directly to (∂l/∂μ_B)... but I don't see the multiplication by 1/m that goes after that term.
 * I'm not sure what `var % arma::mean(-2 * inputMean, 1) / input.n_cols` refers to.

I'm pretty sure I've just misunderstood here; since it passes all the tests, my gut feeling is that it's right.  It might be helpful to add comments describing the conversion from the paper to the code.